## Desafio - Módulo 2 - Machine Learning
##### Respotas do questionário está no final do arquivo

Dataset: https://archive.ics.uci.edu/ml/datasets/wine

**Wine Quality Data Set**<br>
Description

Forina, M. et al, PARVUS -
An Extendible Package for Data Exploration, Classification and Correlation.
Institute of Pharmaceutical and Food Analysis and Technologies, Via Brigata Salerno,
16147 Genoa, Italy.


Attribute Information:

These data are the results of a chemical analysis of wines grown in the same region in Italy but derived from three different cultivars. The analysis determined the quantities of 13 constituents found in each of the three types of wines.

I think that the initial data set had around 30 variables, but for some reason I only have the 13 dimensional version. I had a list of what the 30 or so variables were, but a.) I lost it, and b.), I would not know which 13 variables are included in the set.

The attributes are (dontated by Riccardo Leardi, riclea '@' anchem.unige.it )
- 1) Alcohol
- 2) Malic acid
- 3) Ash
- 4) Alcalinity of ash
- 5) Magnesium
- 6) Total phenols
- 7) Flavanoids
- 8) Nonflavanoid phenols
- 9) Proanthocyanins
- 10)Color intensity
- 11)Hue
- 12)OD280/OD315 of diluted wines
- 13)Proline


### Data Exploration & Analysis

Provide data set splits for training, validation, and testing:

#### Para a implementação dos algoritmos utilize as definições abaixo:
#### Algoritmo KNN: 
clf_KNN = KNeighborsClassifier(n_neighbors=5)
#### Algoritmo Árvore de Decisão: 
clf_arvore = DecisionTreeClassifier()
#### Algoritmo Floresta Randômica: 
clf_floresta = RandomForestClassifier(max_depth=10, random_state=1)
#### Algoritmo SVM: 
clf_svm=SVC(gamma='auto',kernel='rbf')
#### Algoritmo Rede MLP: 
clf_mlp = MLPClassifier( alpha=1e-5, hidden_layer_sizes=(5, 5), random_state=1)
- 1 Quando for realizar a leitura do arquivo “winequality-red.csv” com a função pandas.read_csv() é necessário utilizar o atributo “ sep=';' ” para que as colunas sejam reconhecidas.
- 2 Para a divisão dos dados de treinamento e teste dos algoritmos utilize o valor de “random_state=1” e a proporção de 70% para treinamento e 30% para teste.
- 3 Utilize a normalização dos dados utilizando o MinMaxScaler para todos os algoritmos.
- 4 Utilize a variável “quality” como saída e as demais como entrada do modelo.
- 5 Para a última questão, considere a realização das mesmas etapas desenvolvidas (MinMaxScaler, train_test_split etc).

In [1]:
# importando as bibliotecas
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 

In [2]:
#realizando a leitura do arquivo (dataset)
data = 'winequality-red.csv'
df = pd.read_csv(data, sep=';')

In [3]:
print(list(df.columns))

['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']


In [4]:
#apresentando as 5 primeiras linhas do dataset
df.head()

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

#### Pergunta 1
No dataset utilizado para o desafio, quantas instâncias e atributos existem, respectivamente?

- (12,1599).

- (1345,15).

- ***(1599,12).***

- (15,1345).

In [5]:
df.shape

(1599, 12)

#### Pergunta 2
Quantos tipos diferentes de dados existem no dataset do desafio?

- 17.

- ***2.***

- 12.

- 3.

In [7]:
df.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


#### Pergunta 3
Qual é a proporção (em %) de valores nulos existente na coluna "residual sugar" (resíduo de açucar)?

- 5%.

- 15%.

- 0%.

- 12%.

#### Pergunta 4
Qual é o desvio padrão para a variável "fixed acidity" ?

- 8.3

- 2.3

- 1.7

- 1.4

#### Pergunta 5
Qual é a mediana para a variável "residual sugar"?

- 1.9

- 8.1

- 2.5

- 2.2

#### Pergunta 6
Sobre o coeficiente de correlação de Pearson entre as variáveis "fixed acidity" e "pH" é CORRETO afirmar:

- Existe uma correlação negativa. Quanto maior o "fixed acidity, menor é o "pH".

- Existe uma correlação positiva. Quanto maior o "fixed acidity, menor é o "pH".

- Existe uma correlação negativa. Quanto maior o "fixed acidity, maior é o "pH".

- Existe uma correlação negativa. Quanto menor o "fixed acidity, menor é o "pH".
 

#### Pergunta 7
Mesmo sendo reduzido, é CORRETO dizer sobre o coeficiente de correlação de Pearson entre as variáveis "quality" e "alcohol":

- Para analisar correlação é necessário que o coeficiente seja superior a 1.

- Quanto maior o teor alcoolico (alcohol), maior é a qualidade do vinho (quality).

- Não é possível dizer nada sobre a correlação entre as vairáveis.

- Quanto maior o teor alcoolico (alcohol), menor é a qualidade do vinho (quality).

#### Pergunta 8
Quantas instâncias existem para a qualidade do vinho igual a 5?

- 345

- 681

- 53

- 638

#### Pergunta 9
Aplique o modelo de normalização MinMaxScaler com os valores "default" para os dados de entrada. Qual é o menor valor para a variável "fixed acidity"?

- 1

- 0

- 15.2

- 4,6

#### Pergunta 10
Após dividir o dataset utilizando o método train_test_split(entradas_normalizadas, saida, test_size=0.30, random_state=1). Qual foi a acurária para o KNN?

- 0,56

- 1,2

- 0,15

- 0,89

#### Pergunta 11
Aplicando o algoritmo de árvore de decisão com os parâmetros definidos no enunciado do desafio e a divisão de treinamento e teste como mostrado, anteriormente. Qual é o valor da acurácia?

- 0.80

- -0.58

- 0.58

- -0.80

#### Pergunta 12
Aplicando o algoritmo de floresta randomica com os mesmos parâmetros apresentados no enunciado e mesmos valores treinamento e teste, qual é o valor da acurácia?

- 0,40

- 0,30

- 0,68

- 1,00

#### Pergunta 13
Aplicando o algoritmo SVM para a classificação da qualidade dos vinhos com os mesmos valores anteriores para treinamento e teste, qual é o valor de acurácia obtido?

- 0,72

- 0,36

- 0,15

- 0,59

#### Pergunta 14
Apicando a rede MLP como mostrado no enunciado e com os mesmos dados de treinamento e teste utilizados, qual é a acurácia do modelo?

- 0,60

- 0,90

- 0,25

- 0,30

#### Pergunta 15
Modifique o dataset para um sistema binário que considere vinhos com nota maior do que 5 (6, 7, 8) sejam bons e menores ou igual a 5 ruins (5, 4, 3). Aplique o modelo da floresta randômica. Qual a acurácia do modelo para esse novo dataset?

- 0,80

- 0,50

- 0,68

- 0,60